### ISSUE TITLE
Plugin maintenance meta-issue: tooling, repositories, etc...

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#9606]
As of today, refactoring of the public API in Coq is needed but still remains hard to do due tooling / external repositories. Even if the `create_overlays` script + Dune support in some plugin does help, it is still not optimal as things tend to go out of sync and the process remains clunky [for example no way to test impact on tree easily]

In particular, whether to use a _monorepos_ for the ML parts of plugins has been mentioned several times. This would mean giving access to that part of the repository to plugin authors [likely using bors to check CI], and would bring the main advantage of consistency. The build time for plugins tends to be quite low these days.

Another interesting point is about deprecation. It is clear that deprecation is not very helpful for our development goals as many changes won't be possible to be implemented in this way. In fact, it makes little sense for plugins in CI to treat deprecation warnings as non-fatal as such functions are removed in the next Coq release thus the work has to be done anyways (see also discussion in https://github.com/ocaml/dune/pull/1850 ) We don't deprecate so much anyways these days, as the big deprecation of duplicated stuff has already happened, so the main reasoning to keep this warning non-fatal is due to poor tooling / high cost of doing an overlay.

What should we do? Should we move to a monorepos? Should be develop stronger tooling for consistency and require plugins authors to implement support for it?

https://github.com/coq/coq/issues/6724 is also related to this discussion.

**Comment by Zimmi48** [comment]
FWIW I'd really like to try (or at least discuss more) the solution I proposed in https://github.com/coq/coq/issues/6724#issuecomment-443550002 rather than the full monorepo approach: the repo of Coq contains a folder for each plugin, the plugins keep their repos, and some tool keeps everything synchronized. See also https://github.com/google/copybara / https://github.com/jsirois/sapling (used by Foursquare).

**Comment by Zimmi48** [comment]
@ejgallego What is the current timeline ocaml/dune#1555, especially now that it has missed 1.7.0? It would make these kinds of refactoring much easier.

**Comment by ejgallego** [comment]
@Zimmi48 hard to say, I am optimistic about having it done before the 8.10 freeze; Dune releases pretty often so that doesn't worry me a lot, however there is still some work to do to implement things properly, and I don't have a lot of free time.

[Note that I am developing Coq + Dune in a separate tree from the PR as to avoid noise to Dune devs].

**Comment by ejgallego** [comment]
> FWIW I'd really like to try (or at least discuss more) the solution I proposed in [#6724 (comment)](https://github.com/coq/coq/issues/6724#issuecomment-443550002) rather than the full monorepo approach: the repo of Coq contains a folder for each plugin, the plugins keep their repos, and some tool keeps everything synchronized. See also https://github.com/google/copybara / https://github.com/jsirois/sapling (used by Foursquare).

Interesting, thanks for the links; I have a lot to catch up, let's hope I have done my homework before the next WG where we can hopefully discuss/chat about this.

Something that worries me (a lot) is to have to rely on external tools for such a basic workflow such as having to do a API commit. This both increases complexity and risk in the basic workflow and the cost shouldn't be underestimated.

**Comment by Zimmi48** [comment]
It's all a matter of tradeoffs. But the monorepo approach has other drawbacks in terms of process which I think would damage accessibility to Coq CI and would go against the efforts that we have made to encourage plugin authors to join. These human aspects must not be neglected. BTW, note that the monorepo approach is [not the Linux way](https://blog.ffwll.ch/2017/08/github-why-cant-host-the-kernel.html#this-looks-like-a-monorepo-horror-show).

**Comment by ejgallego** [comment]
Indeed @zimmi48 you are right I am not very familiar with the terminology and I write monorepos when in many cases I may mean monotree. I don't believe we should restrict Coq's CI to the monotree indeed, removing this feature was never in my mind, and we should be able to add developments that are not in-tree for obvious reasons [I am talking here about the `plugins:` part but of course what about all the `v.` stuff, that will never go in tree]

On the other hand the current reality is that refactoring ML code is a large pain, and we are maintaining quite a few plugins. Having multiple repos creates many problems such as for example the need to have two CI's, etc... in the end, it is hard to beat the advantages `git commit` has.

I am familiar with the Linux model and indeed I am a bit afraid of doing it in Coq given the different context. In the Linux model, driver trees have to follow changes in mainline [and mainline often distributes outdated drivers for a while]. It is the cost of the driver developers to rebase their trees when mainstream changes, and they have to submit pull requests.

I am wary of this as it reminds me of `coq-contribs` and how outdated it got as developers didn't update the contribs; in Linux there is way more pressure to do so even if not always happen. So I dunno.

**Comment by ejgallego** [comment]
Another point that is annoying with the current workflow is the need to open multiple pull request on the multiple repositories that form an overlay. Even so we are lucky most plugins are hosted on github, but on a more diverse ecosystem that could be tricky.

**Comment by Zimmi48** [comment]
>Another point that is annoying with the current workflow is the need to open multiple pull request on the multiple repositories that form an overlay. Even so we are lucky most plugins are hosted on github, but on a more diverse ecosystem that could be tricky.

:+1: Opening these PRs automatically was also one of the objectives I was listing in the [above mentioned comment](https://github.com/coq/coq/issues/6724#issuecomment-443550002). I feel like we can do a pretty low-tech solution to support the full generality of platforms which is: if the platform is not GitHub we can push to a fork on GitHub nonetheless and ping maintainers in the main PR thread to let them know that there is an overlay to review (we might even prefer to do that even in the case where the platform is GitHub, as this avoids the multiplicity of threads, which is a problem because it creates scattered discussion).

Anyways, @ejgallego if you want me to start experimenting with this without waiting for native Coq support to land in Dune, I would do it with ltac2 and equations only to start with. The only thing I need from you is the compositional build support. E.g. `make -f Makefile.ci` would now expect a folder `external/ltac2` to be present in the tree and would use Dune to build it (using the current version of Dune + `coq_dune`).

**Comment by ejgallego** [comment]
Thanks @Zimmi48 , I don't recommend you try that route yet, so far the compositional build is only safe tor the ML parts.

I have however updated my toolchain so `create-overlays.sh` has moved to a more genera `overlays` script with a few more commands, I may add `overlays submit` indeed.

**Comment by ejgallego** [comment]
> I'd really like to try (or at least discuss more) the solution I proposed in [#6724 (comment)](https://github.com/coq/coq/issues/6724#issuecomment-443550002) rather than the full monorepo approach: the repo of Coq contains a folder for each plugin, the plugins keep their repos, and some tool keeps everything synchronized.

After thinking a bit more about this and having done quite a bit of work I am not sure why having a local copy is interesting; my tool doesn't care if a folder comes from a different remote, in fact this makes synchronization easy I think as I just have to tell the yet unnamed too to rebase.

**Comment by Zimmi48** [comment]
Emilio suggested having a discussion related to the integration of Ltac2 in the Coq repository.

My point of view is still a wip but at this point I think we should strive to have a hybrid system where we define three things:

- The packages we test in CI. It is pretty open to everyone who wants to be in it, provided they follow our requirements and recommendations.
- The Coq platform, being all the packages we want to bundle with Coq, and which is a subset of the packages we test in CI. There is an explicit selection based on interest. We would distribute this bundle in various forms (Windows, macOS installer, snap package, Docker image...), but these packages could be provided separately on opam or Linux distributions (but with a meta-package that would still install everything in the platform). What should be included in the Coq platform could be discussed with an external team, such as the editorial board of coq-community (when it finally exists).
- The Coq repository, which can serve as a monorepo for the development of a subset of the packages from the Coq platform. To get in, the maintainers of the package must be OK with our development processes (related to reviewing, merging), OK with sharing a single tracker for issues and PRs with the rest of Coq. The documentation of the package must be integrated in Coq's own documentation (refman + coqdoc). There is a moral engagement of the development team to continue the maintenance if the maintainers of the packages became inactive, and thus we can only accept packages for which the interest of the team is sufficiently high.

We must make it easy to migrate to one status to another, and keeping namespaces and opam packages separate (like we are doing for Ltac2) can help with that.

**Comment by Zimmi48** [comment]
FTR UniCoq is a candidate to be the next package to follow the lead of Ltac2: https://github.com/Mtac2/Mtac2/pull/178#issuecomment-490216562

**Comment by ejgallego** [comment]
Something that I am worried about is about the issue tracker; it seems to me that what github provides falls a bit short; but I dunno what do others think.

**Comment by gares** [comment]
While I like unicoq I think 2 unification algorithms are already too many for Coq, and this adds a third one. So I think we need a concrete plan, including unifall in the picture and eventually mtac2.

Anyway I think the topic of integration in Coq should be discussed in person at the next wg or maybe at the cudw.

**Comment by Zimmi48** [comment]
>There is a moral engagement of the development team to continue the maintenance if the maintainers of the packages became inactive, and thus we can only accept packages for which the interest of the team is sufficiently high.

This is the precondition to integration that makes it natural for @maximedenes and @gares to ask for a transition plan, so that including UniCoq is not just going to create more maintenance burden in the long run.

**Comment by Zimmi48** [comment]
> Something that I am worried about is about the issue tracker; it seems to me that what github provides falls a bit short; but I dunno what do others think.

Your message is a bit vague on what are the limitations of GitHub's issue tracker that worry you. For my part, I clearly see one specific limitation which is related to notifications: having a single issue tracker may create lots of noise for the maintainers of a component or plugin who are specifically interested in following the bug reports about this component or plugin only. In the case of Ltac2, this is not a problem because Pierre-Marie is already following the rest of the issues, but in the case of UniCoq, it could be a problem for @beta-ziliani. But I don't think the problem is new: we already have this problem for maintainers of plugins that are currently part of the stdlib (e.g. @amahboubi, @fajb, @forestjulien, @PierreCorbineau, @thery).

I see several solutions:

- We can have other projects dedicated to tracking issues only for some specific plugins, even if their code is in the monorepo: in this case, the maintainers only watch this repo and the developers that watch the Coq repo transfer any misplaced issue from Coq to the specific tracker. I'm not very fond of this solution, but it has the advantage of providing a simple way, with no new technical requirement that allows to recreate a division in components. One drawback would be that people who want to follow all the issues would have to watch many repositories, and they probably won't, which could lead to less attention being given to the issues on the other trackers.

- We could find a workflow, based on labels, using existing tools such as Octobox, and teach it to interested maintainers.

- We could @-mention maintainers automatically whenever a label is added (would be easy to add the feature to @coqbot).

Most importantly, we need to know about the expectations of maintainers to help devise the right solution.

**Comment by ejgallego** [comment]
> Your message is a bit vague on what are the limitations of GitHub's issue tracker that worry you.

I didn't think of anything too specific but more of, as you point out, of limitations in terms of restricting the issue view to a particular component. So indeed there should be I guess a recommended set of practices as you have outlined.

**Comment by Zimmi48** [comment]
Restricting the issue view with labels is quite easy though.

**Comment by gares** [comment]
> This is the precondition to integration that makes it natural for @maximedenes and @gares to ask for a transition plan

Yes, but I think we should also define clearly which is a trajectory that is acceptable for an addon.
To me your 3 statuses are in a precise order: you only move downward.

To me the third status should only be reached by code that was validated by users. Hence I'm all for shipping stuff that we have in CI (upon author request) into out installers (this is status 2). But status 3 should be, IMO, something you reach when the add on has a substantial mass of users (that can reach because it was in status 2 for some time).

For example I (privately) questioned the PR of @maximedenes about Ltac2. I'm reassured by the relative simplicity of its code (that I enjoyed reading) and simplicity of its design and by the fact that his author is already an established dev and finally that Coq is terribly in need of something less shaky than Ltac1. Still I'm not very convinced that the software was properly tested and has a reasonable user base, but I got convinced that the benefit in this case was outweighing the risk, so I did not raise any question. Still I hope we start dropping Ltac1 from our stdlib1 (or 2), for example the ring tactic should be ported to Ltac2 (that is indeed a transition plan that was not discussed, but seems the only way to kill Ltac1).

When an add on reaches the third status I expect it to be no more an add on, but something a substantial part of our users depend on, hence there is no going back (eg rewriting ring in Ltac2 makes it no more an optional package). Stuff that is in status 3 can still move, but IMO not back to status 2, but only to the trash can. FTR we tried to move something backward and it did not work well (bignum) simply because it was used. We moved it to status 1, then back to 2 in urgency on some platforms, and still today there is no clear win to me in not having it in status 3.

As a conclusion I believe that moving something to status 3 should be done very carefully and a transition plan is part of this, but also a concrete assessment of risk/benefit and quality. The technical details (CI, BTS, ...) are very minor IMO, or at least do not worry me. The moral engagement is a bit idealistic and not time proof to me (there is plenty of code we are maintaining that we did not sign for, and I find it quite inevitable).

**Comment by beta-ziliani** [comment]
Hi, as a plugin dev, I want to point out two things: 1) I agree with @gares that the technical details are unimportant, and @Zimmi48 's labeling system + mentions sounds good to me. 2) More importantly, about when to incorporate a plugin into stage 3 (as @gares puts it), I guess there is a coq-egg problem: the number of users might increase once it reaches stage 3. I'm certain Ltac2 will have users once it is "ready-to-use" in Coq, and the same goes to Unicoq (but in a much smaller scale).

**Comment by Zimmi48** [comment]
I agree with both of you. The trajectory proposed by @gares is the ideal one. However, as noted by @beta-ziliani, the statuses that precede stage 3 may not be suitable to gain enough users. In fact, it is very dependent on the non-technical effort put by developers of said add-ons / plugins. For instance, Equations seems to have reached quite many users, because Matthieu put a lot of effort in releasing often and advertising new releases. OTOH Ltac2 always struggled to gather users, but they were often not even aware that they could install it and use it. The question we should ask ourselves is whether it is normal that the target of an add-on will depend on these aspects, whether we are fine doing nothing more about it, whether we want to devise a set of good practices that are easy to follow and help make sure packages can gain users, or whether we want to go beyond and take care of advertising the packages that look interesting to us so that they gain users, even if their authors might not know what to do about it.

**Comment by gares** [comment]
>  I guess there is a coq-egg problem: the number of users might increase once it reaches stage 3.

> However, as noted by @beta-ziliani, the statuses that precede stage 3 may not be suitable to gain enough users.

Can you two elaborate on the reasons why stage 2 would not be sufficient?

Unless I'm mistaken, an addon in stage 2 is accessible by `Require foo` *on any Coq install* (it is part of the bundles, and we can make the Coq opam package just depend on these, for example). The main difference is, to me, that something in stage 2 might go away in the next version of Coq, so you may not want to use it "in production", while something in stage 3 is going to stay unless a better replacement is found.

Being is stage 2 requires some coordination "at release time" like tagging the version one wants to be released (that should also be the one being tested by CI) and providing an extract for the ChangeLog. So to me it should ease, from the addon dev POV, the release process at the cost of some temporal synchronization with the releases of Coq. I would personally consider this OK for coq-elpi.

**Comment by beta-ziliani** [comment]
The only difference would then be where the repo is, right? I think the reason behind moving plugins in 2 to 3 is to ease maintenance (for you and us), since releases must be coordinated.

**Comment by Zimmi48** [comment]
Indeed, @gares, I agree that stage 2, if we make such an effort towards providing good bundles on all platforms, with enough coordination at release time, should probably be sufficient to solve the chicken-and-egg problem. This is more-or-less what I meant with my questions above, but I think you expressed well what would be needed.

@beta-ziliani It depends. I had initially written "The documentation of the package must be integrated in Coq's own documentation (refman + coqdoc). There is a moral engagement of the development team to continue the maintenance if the maintainers of the packages became inactive, and thus we can only accept packages for which the interest of the team is sufficiently high.", but maybe there should be a stage 3 and 4, and 3 should not go as far but being, as you said, only a matter of where the repo is.

**Comment by gares** [comment]
You are right about that, no overlays would be needed.
Still this should not impact the number of users an addon can gather by being in stage 2 and not 3 (that was my original question).

**Comment by beta-ziliani** [comment]
sure, I was just pointing out what could be the gain. stage 2 to me (as plugin dev) is already a big win.
